FABN-1224 NodeSDK discovery interest 1.4 (#132)
hyperledger/fabric-sdk-node: docs/tutorials/discovery-fabric-network.md

This tutorial illustrates the use of the service discovery by the Hyperledger Fabric Node.js fabric-network SDK.

For more information on:
* [Service Discovery](https://hyperledger-fabric.readthedocs.io/en/release-1.4/discovery-overview.html#how-service-discovery-works-in-fabric)

The following assumes an understanding of the Hyperledger Fabric network
(orderers and peers),
and of Node application development, including the use of the
Javascript `promise` and `async await`.

### Overview

> The service discovery provided by the Hyperledger Fabric helps an application to

Author: fixed

> understand the current view of the network. The service discovery also has insight

Author: fixed

> into the endorsement policies of chaincodes and is able to provide various list of

Author: fixed

peers that are currently active on the network that could be used to endorse a
proposal.

> To use the discovery service the application will have to define just one peer.

Author: fixed

> #### fabric-network API's that will use the service discovery

Author: fixed

local system; otherwise should be disabled.

### To use
By default the fabric-network will not use the service discovery. To enable the
use of discovery, set the discovery 'enabled' attribute to a value of true.

```
await gateway.connect(connectionProfile, {discovery: { enabled: true, asLocalhost: false}});
const network = await gateway.getNetork('mychannel');

```

### To use with docker-compose
When the fabric network is running in a docker-compose and the node.js application
is running outside of the docker containers, it will be necessary to modify the
addresses returned from the discovery service. The discovery service sees the
addresses of the peers and orderers as host names and ports as they exist in the
virtual systems, however the node.js
application running outside of docker will only know the endpoints as `localhost`
and port. In the docker-compose file, notice how Docker is mapping the port addresses,
these must be same when using service discovery. Using `- 7061:7051` will not
work as the application does not have visibility into the virtual system and
the port address as seen by the peer's discovery service.

> Use the `asLocalhost` with true or false, the default is true.

Reviewer: ```suggestion
Set the `asLocalhost` attribute to `true`.
```

```
await gateway.connect(connectionProfile, {discovery: { enabled: true, asLocalhost: true}});
const network = await gateway.getNetork('mychannel');
```

> Notice in the following definition of a peer from a docker-compose file.

Reviewer: ```suggestion
Notice in the following definition of a peer from a docker-compose.yaml file that
```

> The port number is the same and has been defined along with the

Reviewer: ```suggestion
the port number is the same and has been defined along with the
```

host name `peer0.org1.example.com:7051` for the peer and gossip settings.

> The node.js fabric-client application running outside of of the docker

Reviewer: ```suggestion
The Node.js fabric-client application running outside of of the Docker
```

> containers will use `localhost:7051`. There is a mapping of the host name to

Reviewer: ```suggestion
network will use `localhost:7051`. There is a mapping of the host name to
```

> 'localhost' but there is no mapping of the port address.

Reviewer: ```suggestion
'localhost', but there is no mapping of the port address.
```

	- FABRIC_LOGGING_SPEC=debug
	## the following setting redirects chaincode container logs to the peer container logs
	- CORE_VM_DOCKER_ATTACHSTDOUT=true
	- CORE_PEER_LOCALMSPID=Org1MSP
	- CORE_PEER_MSPCONFIGPATH=/etc/hyperledger/msp/peer
	##
	- CORE_PEER_TLS_ENABLED=true
	- CORE_PEER_TLS_CLIENTAUTHREQUIRED=true
	- CORE_PEER_TLS_KEY_FILE=/etc/hyperledger/msp/peer/tls/key.pem
	- CORE_PEER_TLS_CERT_FILE=/etc/hyperledger/msp/peer/tls/cert.pem
	- CORE_PEER_TLS_ROOTCERT_FILE=/etc/hyperledger/msp/peer/cacerts/org1.example.com-cert.pem
	- CORE_PEER_TLS_CLIENTROOTCAS_FILES=/etc/hyperledger/msp/peer/cacerts/org1.example.com-cert.pem
	# # the following setting starts chaincode containers on the same
	# # bridge network as the peers
	# # https://docs.docker.com/compose/networking/
	- CORE_VM_DOCKER_HOSTCONFIG_NETWORKMODE=fixtures_default
  working_dir: /opt/gopath/src/github.com/hyperledger/fabric
  command: peer node start
  ports:
	- 7051:7051
  volumes:
	  - /var/run/:/host/var/run/
	  - ./channel/crypto-config/peerOrganizations/org1.example.com/peers/peer0.org1.example.com/:/etc/hyperledger/msp/peer
  depends_on:
	- orderer.example.com
```

> ### using chaincode to chaincode calls and collections

Reviewer: ```suggestion
### Using chaincode to chaincode calls and collections
```

> The peer's discovery service will required information

Reviewer: ```suggestion
The peer's discovery service requires information
```

on the channel, chaincodes, and collections that are involved in the

> smart contract to develop a valid endorsement plan.

Reviewer: ```suggestion
smart contract in order to develop a valid endorsement plan.
```

The contract's channel name and chaincode name will be automatically sent to
the peer's discovery service unless the contract indicates that it requires
other chaincodes and collections. These are known as 'discovery interests' to the
peer's discovery service.

> If the endorsement will require one or more chaincode to chaincode calls and/or

Reviewer: ```suggestion
If the endorsement requires one or more chaincode to chaincode calls and/or
```

> be over collections, then the application must tell the contract object

Reviewer: ```suggestion
collections, then the application must inform the contract object
```

instance of these names. This will assist the discovery service in putting

> together an endorsement plan based on all the endorsement policies of chaincodes

Reviewer: ```suggestion
together an endorsement plan based on all the endorsement policies of the chaincodes
```

Author: fixed

and collections involved and the active peers on the network.
The endorsement plan will include those peers that satisfy the endorsement
policies and the collection policies.

> The following examples shows how to setup the interest JSON object to

Reviewer: ```suggestion
The following examples show how to setup the interest JSON object to
```

Author: fixed

provide to the contract instance when
a chaincode to chaincode call over collections is required.
The discovery interest will used as is, therefore the contract's chaincode name
should be included along with any collections.

```
// use the following when the smart contract's chaincode will be
// calling another chaincode
const interest_chaincodes = {
	chaincodes: [
		{name: 'mychaincode1'},
		{name: 'mychaincode2'}
	]
};

// use the following when the smart contract's chaincode is in a
// collection
const interest_chaincode_collection = {
	chaincodes: [
		{
			name: 'mychaincode1',
			collection_names: ['mycollection1']
		}
	]
};

> // use the following when the smart contract's chaincode in multiple

Reviewer: ```suggestion
// use the following when the smart contract's chaincode is in multiple
```

Author: fixed
